### ISSUE TITLE
Provide better error message when Amplify.addPlugin is called before runApp()

### ISSUE THREAD

**Issue by Jordan-Nelson** [aws-amplify/amplify-flutter#1309]
**Context**

A customer shared via discord that they were receiving a null pointer exception when configuring amplify. This is because Amplify.addPlugin() was being invoked from `main()` before `runApp()` was called. See code snippet below.

```dart
Future<void> main() async {
  await _configureAmplify(); // calls addPlugin() and configure()
  runApp(MyApp());
} 
```

This throws an exception that isn't very helpful. The exception is the result of the binaryMessenger (which is used by the method channel) not yet being available. The method channel unfortunately assumes this is available, casting it to non-null.

Although it is not what we recommend in the docs, this _can be_ resolved by adding a call to `WidgetsFlutterBinding.ensureInitialized();`. See code snippet below.

```dart
Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await _configureAmplify(); // calls addPlugin() and configure()
  runApp(MyApp());
} 
```

**Request**

Throw a more helpful exception in this instance by checking is the binaryMessenger is available before attempting to invoke a method. Something like the code below would achieve this:

```dart
Future<T?> safeInvokeMethod<T>(
  MethodChannel channel,
  String method, [
  dynamic arguments,
]) {
  try {
    channel.binaryMessenger;
  } catch (e) {
    throw AmplifyException(
      'You are attempting to configure amplify before calling runApp().',
    );
  }
  return channel.invokeMethod(method, arguments);
}
```

**Discord Chat Details**

<details>

Sal — Yesterday at 7:15 AM

> i tried to move the confiugAmplify method into the main class
and now i get this:
I/flutter (22103): Amplify plugin was not added
E/flutter (22103): [ERROR:flutter/lib/ui/ui_dart_state.cc(209)] Unhandled Exception: AmplifyException(message: Amplify plugin AmplifyAuthCognito was not added successfully., recoverySuggestion: We currently don't have a recovery suggestion for this exception., underlyingException: Null check

Sal — Yesterday at 7:16 AM

> This is the method:
Future<void> _configureAmplify() async {
  AmplifyAuthCognito authPlugin = AmplifyAuthCognito();
  await Amplify.addPlugins([authPlugin]);
  try {
    await Amplify.configure(amplifyconfig);
  } on AmplifyAlreadyConfiguredException {
    print(
        "Tried to reconfigure Amplify; this can occur when your app restarts on Android.");
  }
} 

jordan nelson — Yesterday at 2:18 PM

> Hi @Sal - The code to configure amplify looks correct to me. It may help if you shared your amplify config (with sensitive info removed) and what platform (iOS or Android) you are running on. You can also create a github issue with that info and we can try to reproduce the issue you are facing.

Sal — Today at 9:25 AM

> Hi Jordan, here it is:
const amplifyconfig = ''' {
    "UserAgent": "aws-amplify-cli/2.0",
    "Version": "1.0",
    "auth": {
        "plugins": {
            "awsCognitoAuthPlugin": {
                "UserAgent": "aws-amplify-cli/0.1.0",
                "Version": "0.1.0",
                "IdentityManager": {
                    "Default": {}
                },
                "CognitoUserPool": {
                    "Default": {
                        "PoolId": "xxx",
                        "AppClientId": "xxx",
                        "Region": "xxx"
                    }
                },
                "Auth": {
                    "Default": {
                        "authenticationFlowType": "USER_SRP_AUTH"
                    }
                }
            }
        }
    }
    }
}''';
I'm running it on an android emulator

jordan nelson — Today at 9:49 AM

> Was that config generated with the amplify CLI or hand edited?

jordan nelson — Today at 9:56 AM

> Can you let me know what version of amplify_flutter you are using, and what version of Android OS the simulator is running?

jordan nelson — Today at 10:06 AM

> Never mind. I was able to reproduce this with your code snippets. The issue is that you are calling _configureAmplify() in main(). I believe all of our docs should have this called inside an init state method in a stateful widget. If you see an example where this is called inside of main(), let me know.
>
> There are two option to resolve this.
>
> 1. If you want to keep the call to  _configureAmplify() in main, that you need to call WidgetsFlutterBinding.ensureInitialized(); before you call _configureAmplify(). I wouldn't really recommend this unless you had a need to call it in main.
>
> 2. Move the call to _configureAmplify() into a stateful widget's init state method. The getting started guide has an example of this. This is what I would recommend. See https://docs.amplify.aws/lib/auth/getting-started/q/platform/flutter/#initialize-amplify-auth

</details>

**Other Info / Options**

There is [an open issue with Flutter](https://github.com/flutter/flutter/issues/78028) to provide a better exception and/or message. 

If this were resolved, we may not need to do anything. At the very least, we could just catch the exception if a specific exception was thrown. Checking if the binaryMessenger is available requires knowing some of the internal details of how a MethodChannel works. Checking for the existence of the binaryMessenger would couple amplify-flutter to that internal implementation. [Here is the location where the non-null cast happens](https://github.com/flutter/flutter/blob/10658266866d4d7e88b5e3b9e435f5a06273019f/packages/flutter/lib/src/services/platform_channel.dart#L52).

**Comment by Jordan-Nelson** [comment]
https://github.com/flutter/flutter/pull/92758 added a better error message back into master. https://github.com/flutter/flutter/pull/89451 (as is) cleans up some of the error messages and will add a `checkInstance` method, which I believe could be used if we wanted to catch this as provide and even better error message.

I'll look to test this out with those changes in the future, and if those error messages seem sufficient I will close this out.

**Comment by Jordan-Nelson** [comment]
The PRs above have been merged into flutter main, and in flutter 2.10.3 the message is more helpful.

```
Unhandled Exception: 'package:flutter[/src/services/platform_channel.dart]()': Failed assertion: line 134 pos 7: '_binaryMessenger != null || ServicesBinding.instance != null': Cannot use this MethodChannel before the binary messenger has been initialized. This happens when you invoke platform methods before the WidgetsFlutterBinding has been initialized. You can fix this by either calling WidgetsFlutterBinding.ensureInitialized() before this or by passing a custom BinaryMessenger instance to MethodChannel().
```

This message could still be a little confusing to a user on Amplify-Flutter, but don't think it is worth trying to catch this just to throw an error with a slightly different message.

Closing this out.
